File: smart-toast/src/main/java/com/coder/zzq/smartshow/toast/emotion/EmotionToastInvoker.java
package com.coder.zzq.smartshow.toast.emotion;

import android.view.Gravity;
import android.widget.Toast;

import com.coder.zzq.smartshow.toast.schedule.ToastScheduler;
import com.coder.zzq.toolkit.Utils;

public class EmotionToastInvoker implements EmotionToastView.Overall, EmotionToastView.ConfigSetter {
    private final EmotionToast.Config mConfig = new EmotionToast.Config();

    @Override
    public EmotionToastView.ConfigSetter config() {
        return this;
    }

    @Override
    public EmotionToastView.ConfigSetter cancelOnActivityExit(boolean b) {
        mConfig.mCancelOnActivityExit = b;
        return this;
    }

    @Override
    public EmotionToastView.ConfigSetter backgroundColor(int color) {
        mConfig.mBackgroundColor = color;
        return this;
    }

    @Override
    public EmotionToastView.ConfigSetter backgroundColorResource(int colorResource) {
        return backgroundColor(Utils.getColorFromRes(colorResource));
    }

    @Override
    public EmotionToastView.ConfigSetter iconResource(int icon) {
        mConfig.mIcon = icon;
        return this;
    }

    @Override
    public EmotionToastView.ConfigSetter iconSizeDp(float sizeDp) {
        mConfig.mIconSize = Utils.dpToPx(sizeDp);
        return this;
    }

    @Override
    public EmotionToastView.ConfigSetter iconPaddingDp(float padding) {
        mConfig.mIconPadding = Utils.dpToPx(padding);
        return this;
    }

    @Override
    public EmotionToastView.ConfigSetter messageColor(int color) {
        mConfig.mMsgColor = color;
        return this;
    }

    @Override
    public EmotionToastView.ConfigSetter messageColorResource(int colorResource) {
        return messageColor(Utils.getColorFromRes(colorResource));
    }

    @Override
    public EmotionToastView.ConfigSetter messageSizeSp(float sizeSp) {
        mConfig.mMsgSizeSp = sizeSp;
        return this;
    }

    @Override
    public EmotionToastView.ConfigSetter messageBold(boolean bold) {
        mConfig.mMsgBold = bold;
        return this;
    }

    @Override
    public EmotionToastApi apply() {
        return this;
    }

    @Override
    public void info(CharSequence msg) {
        showHelper(EmotionToast.Config.EMOTION_TYPE_INFO, msg, Toast.LENGTH_SHORT);
    }

    @Override
    public void info(int msg) {
        info(Utils.getStringFromRes(msg));
    }

    @Override
    public void infoLong(CharSequence msg) {
        showHelper(EmotionToast.Config.EMOTION_TYPE_INFO, msg, Toast.LENGTH_LONG);
    }

    @Override
    public void infoLong(int msg) {
        infoLong(Utils.getStringFromRes(msg));
    }

    @Override
    public void warning(CharSequence msg) {
        showHelper(EmotionToast.Config.EMOTION_TYPE_WARNING, msg, Toast.LENGTH_SHORT);
    }

    @Override
    public void warning(int msg) {
        warning(Utils.getStringFromRes(msg));
    }

    @Override
    public void warningLong(CharSequence msg) {
        showHelper(EmotionToast.Config.EMOTION_TYPE_WARNING, msg, Toast.LENGTH_LONG);
    }

    @Override
    public void warningLong(int msg) {
        warningLong(msg);
    }

    @Override
    public void success(CharSequence msg) {
        showHelper(EmotionToast.Config.EMOTION_TYPE_SUCCESS, msg, Toast.LENGTH_SHORT);
    }

    @Override
    public void success(int msg) {
        success(Utils.getStringFromRes(msg));
    }

    @Override
    public void successLong(CharSequence msg) {
        showHelper(EmotionToast.Config.EMOTION_TYPE_SUCCESS, msg, Toast.LENGTH_LONG);
    }

    @Override
    public void successLong(int msg) {
        successLong(Utils.getStringFromRes(msg));
    }

    @Override
    public void error(CharSequence msg) {
        showHelper(EmotionToast.Config.EMOTION_TYPE_ERROR, msg, Toast.LENGTH_SHORT);
    }

    @Override
    public void error(int msg) {
        error(Utils.getStringFromRes(msg));
    }

    @Override
    public void errorLong(CharSequence msg) {
        showHelper(EmotionToast.Config.EMOTION_TYPE_ERROR, msg, Toast.LENGTH_LONG);
    }

    @Override
    public void errorLong(int msg) {
        errorLong(Utils.getStringFromRes(msg));
    }

    @Override
    public void fail(CharSequence msg) {
        showHelper(EmotionToast.Config.EMOTION_TYPE_FAIL, msg, Toast.LENGTH_SHORT);
    }

    @Override
    public void fail(int msg) {
        fail(Utils.getStringFromRes(msg));
    }

    @Override
    public void failLong(CharSequence msg) {
        showHelper(EmotionToast.Config.EMOTION_TYPE_FAIL, msg, Toast.LENGTH_LONG);
    }

    @Override
    public void failLong(int msg) {
        failLong(Utils.getStringFromRes(msg));
    }

    @Override
    public void complete(CharSequence msg) {
        showHelper(EmotionToast.Config.EMOTION_TYPE_COMPLETE, msg, Toast.LENGTH_SHORT);
    }

    @Override
    public void complete(int msg) {
        complete(Utils.getStringFromRes(msg));
    }

    @Override
    public void completeLong(CharSequence msg) {
        showHelper(EmotionToast.Config.EMOTION_TYPE_COMPLETE, msg, Toast.LENGTH_LONG);
    }

    @Override
    public void completeLong(int msg) {
        completeLong(Utils.getStringFromRes(msg));
    }

    @Override
    public void forbid(CharSequence msg) {
        showHelper(EmotionToast.Config.EMOTION_TYPE_FORBID, msg, Toast.LENGTH_SHORT);
    }

    @Override
    public void forbid(int msg) {
        forbid(Utils.getStringFromRes(msg));
    }

    @Override
    public void forbidLong(CharSequence msg) {
        showHelper(EmotionToast.Config.EMOTION_TYPE_FORBID, msg, Toast.LENGTH_LONG);
    }

    @Override
    public void forbidLong(int msg) {
        forbidLong(Utils.getStringFromRes(msg));
    }

    @Override
    public void waiting(CharSequence msg) {
        showHelper(EmotionToast.Config.EMOTION_TYPE_WAITING, msg, Toast.LENGTH_SHORT);
    }

    @Override
    public void waiting(int msg) {
        waiting(Utils.getStringFromRes(msg));
    }

    @Override
    public void waitingLong(CharSequence msg) {
        showHelper(EmotionToast.Config.EMOTION_TYPE_WAITING, msg, Toast.LENGTH_LONG);
    }

    @Override
    public void waitingLong(int msg) {
        waitingLong(Utils.getStringFromRes(msg));
    }

    public void showHelper(@EmotionToast.Config.EmotionType int emotionType, CharSequence msg, int duration) {
        mConfig.mEmotionType = emotionType;
        mConfig.mMsg = msg;
        mConfig.mDuration = duration;
        mConfig.mGravity = Gravity.CENTER;
        mConfig.mXOffset = 0;
        mConfig.mYOffset = 0;
        ToastScheduler.get().schedule(EmotionToastFactory.get().produceToast(mConfig));
    }
}

File: smart-toast/src/main/java/com/coder/vincent/smart_toast/alias/emotion/EmotionToastInvoker.kt
package com.coder.vincent.smart_toast.alias.emotion

import android.graphics.drawable.Drawable
import android.view.Gravity
import androidx.annotation.ColorInt
import androidx.annotation.ColorRes
import androidx.annotation.DrawableRes
import androidx.annotation.StringRes
import com.coder.vincent.series.common_lib.bean.TextStyle
import com.coder.vincent.series.common_lib.resourceToColor
import com.coder.vincent.series.common_lib.resourceToDrawable
import com.coder.vincent.series.common_lib.resourceToString
import com.coder.vincent.smart_toast.EMOTION_COMPLETE
import com.coder.vincent.smart_toast.EMOTION_ERROR
import com.coder.vincent.smart_toast.EMOTION_FAIL
import com.coder.vincent.smart_toast.EMOTION_FORBID
import com.coder.vincent.smart_toast.EMOTION_INFO
import com.coder.vincent.smart_toast.EMOTION_SUCCESS
import com.coder.vincent.smart_toast.EMOTION_WAITING
import com.coder.vincent.smart_toast.EMOTION_WARNING
import com.coder.vincent.smart_toast.ShowEmotionToast
import com.coder.vincent.smart_toast.TOAST_ALIAS_EMOTION
import com.coder.vincent.smart_toast.bean.Duration
import com.coder.vincent.smart_toast.factory.Location
import com.coder.vincent.smart_toast.schedule.ToastScheduler

class EmotionToastInvoker : EmotionToastFacade.Overall, EmotionToastFacade.ConfigSetter {
    private val config: EmotionToast.Config =
        EmotionToast.Config().apply { alias = TOAST_ALIAS_EMOTION }

    override fun config(): EmotionToastFacade.ConfigSetter = this

    override fun info(msg: CharSequence) {
        showHelper(EMOTION_INFO, msg, config.duration)
    }

    override fun info(@StringRes msg: Int) {
        info(msg.resourceToString())
    }

    @Deprecated("use config().duration(Duration.LONG) instead.")
    override fun infoLong(msg: CharSequence) {
        showHelper(EMOTION_INFO, msg, Duration.LONG)
    }

    @Deprecated("use config().duration(Duration.LONG) instead.")
    override fun infoLong(@StringRes msg: Int) {
        info(msg.resourceToString())
    }

    override fun warning(msg: CharSequence) {
        showHelper(EMOTION_WARNING, msg, config.duration)
    }

    override fun warning(@StringRes msg: Int) {
        warning(msg.resourceToString())
    }

    @Deprecated("use config().duration(Duration.LONG) instead.")
    override fun warningLong(msg: CharSequence) {
        showHelper(EMOTION_WARNING, msg, Duration.LONG)
    }

    @Deprecated("use config().duration(Duration.LONG) instead.")
    override fun warningLong(@StringRes msg: Int) {
        warning(msg.resourceToString())
    }

    override fun success(msg: CharSequence) {
        showHelper(EMOTION_SUCCESS, msg, config.duration)
    }

    override fun success(@StringRes msg: Int) {
        success(msg.resourceToString())
    }

    @Deprecated("use config().duration(Duration.LONG) instead.")
    override fun successLong(msg: CharSequence) {
        showHelper(EMOTION_SUCCESS, msg, Duration.LONG)
    }

    @Deprecated("use config().duration(Duration.LONG) instead.")
    override fun successLong(@StringRes msg: Int) {
        success(msg.resourceToString())
    }

    override fun error(msg: CharSequence) {
        showHelper(EMOTION_ERROR, msg, config.duration)
    }

    override fun error(@StringRes msg: Int) {
        error(msg.resourceToString())
    }

    @Deprecated("use config().duration(Duration.LONG) instead.")
    override fun errorLong(msg: CharSequence) {
        showHelper(EMOTION_ERROR, msg, Duration.LONG)
    }

    @Deprecated("use config().duration(Duration.LONG) instead.")
    override fun errorLong(@StringRes msg: Int) {
        error(msg.resourceToString())
    }

    override fun fail(msg: CharSequence) {
        showHelper(EMOTION_FAIL, msg, config.duration)
    }

    override fun fail(@StringRes msg: Int) {
        fail(msg.resourceToString())
    }

    @Deprecated("use config().duration(Duration.LONG) instead.")
    override fun failLong(msg: CharSequence) {
        showHelper(EMOTION_FAIL, msg, Duration.LONG)
    }

    @Deprecated("use config().duration(Duration.LONG) instead.")
    override fun failLong(@StringRes msg: Int) {
        fail(msg.resourceToString())
    }

    override fun complete(msg: CharSequence) {
        showHelper(EMOTION_COMPLETE, msg, config.duration)
    }

    override fun complete(@StringRes msg: Int) {
        complete(msg.resourceToString())
    }

    @Deprecated("use config().duration(Duration.LONG) instead.")
    override fun completeLong(msg: CharSequence) {
        showHelper(EMOTION_COMPLETE, msg, Duration.LONG)
    }

    @Deprecated("use config().duration(Duration.LONG) instead.")
    override fun completeLong(@StringRes msg: Int) {
        complete(msg.resourceToString())
    }

    override fun forbid(msg: CharSequence) {
        showHelper(EMOTION_FORBID, msg, config.duration)
    }

    override fun forbid(@StringRes msg: Int) {
        forbid(msg.resourceToString())
    }

    @Deprecated("use config().duration(Duration.LONG) instead.")
    override fun forbidLong(msg: CharSequence) {
        showHelper(EMOTION_FORBID, msg, Duration.LONG)
    }

    @Deprecated("use config().duration(Duration.LONG) instead.")
    override fun forbidLong(@StringRes msg: Int) {
        forbid(msg.resourceToString())
    }

    override fun waiting(msg: CharSequence) {
        showHelper(EMOTION_WAITING, msg, config.duration)
    }

    override fun waiting(@StringRes msg: Int) {
        waiting(msg.resourceToString())
    }

    @Deprecated("use config().duration(Duration.LONG) instead.")
    override fun waitingLong(msg: CharSequence) {
        showHelper(EMOTION_WAITING, msg, Duration.LONG)
    }

    @Deprecated("use config().duration(Duration.LONG) instead.")
    override fun waitingLong(@StringRes msg: Int) {
        waiting(msg.resourceToString())
    }

    private fun showHelper(
        emotion: Int,
        msg: CharSequence,
        duration: Duration,
    ) {
        config.emotion = emotion
        config.message = msg
        config.duration = duration
        config.location = Location(Gravity.CENTER, 0, 0)
        ToastScheduler.schedule(config, EmotionToastFactory)
    }

    override fun backgroundDrawable(bgDrawable: Drawable): EmotionToastFacade.ConfigSetter =
        this.apply {
            config.backgroundDrawable = bgDrawable
        }

    override fun backgroundResource(@DrawableRes bgResource: Int): EmotionToastFacade.ConfigSetter =
        this.apply {
            config.backgroundDrawable = bgResource.resourceToDrawable()!!
        }

    override fun messageStyle(
        @ColorInt color: Int,
        size: Float,
        bold: Boolean,
    ): EmotionToastFacade.ConfigSetter = this.apply {
        config.messageStyle = TextStyle(color, size, bold)
    }

    override fun iconDrawable(iconDrawable: Drawable?): EmotionToastFacade.ConfigSetter =
        this.apply {
            config.iconDrawable = iconDrawable
        }

    override fun iconResource(@DrawableRes iconResource: Int): EmotionToastFacade.ConfigSetter =
        this.apply {
            config.iconDrawable = iconResource.resourceToDrawable()
        }

    override fun iconSize(size: Float): EmotionToastFacade.ConfigSetter = this.apply {
        config.iconSize = size
    }

    override fun marginBetweenIconAndMsg(margin: Float): EmotionToastFacade.ConfigSetter =
        this.apply {
            config.marginBetweenIconAndMsg = margin
        }

    override fun backgroundColor(backgroundColor: Int): EmotionToastFacade.ConfigSetter =
        this.apply {
            config.backgroundColor = backgroundColor
        }

    override fun backgroundColorResource(@ColorRes backgroundColorResource: Int):
            EmotionToastFacade.ConfigSetter = this.apply {
        config.backgroundColor = backgroundColorResource.resourceToColor()
    }

    override fun duration(duration: Duration): EmotionToastFacade.ConfigSetter =
        this.apply {
            config.duration = duration
        }

    override fun commit(): ShowEmotionToast = this
}
